### PR TITLE
Add java test for threads

### PR DESCRIPTION
**Pull request by ilausuch**
https://progress.opensuse.org/issues/110689
VR: http://marvin5.arch.suse.cz/tests/449

### CODE REVIEW

File: tests/test_openjdk.py [hunk 1]
```diff
@@ -28,3 +30,39 @@ def test_jdk_version(container, java_version):
...
+)
+def test_jdk_extended(
+    container_per_test, test_to_run, expected_string, host, container_runtime
+):
+    # copy the test to the container
+    host.run_expect(
+        [0],
+        f"{container_runtime.runner_binary} cp {appdir + test_to_run}.java"
+        f" {container_per_test.container_id}:{condir}",
+    )
```
Comment (dcermak) [inline tests/test_openjdk.py:56]: I would suggest to perform the copy at container build time, as that will reduce the overall test runtime.

File: tests/test_openjdk.py [hunk 2]
```diff
@@ -28,3 +30,39 @@ def test_jdk_version(container, java_version):
...
+    testout = container_per_test.connection.run_expect(
+        [0], "java " + condir + test_to_run + ".java"
+    )
+
+    # check the results
+    if isinstance(expected_string, list):
+        for check in expected_string:
+            assert check in testout.stdout
+    else:
+        assert expected_string in testout.stdout
```
Comment (dcermak) [inline tests/test_openjdk.py:68]: This should not be necessary, you do **know** which type `expected_string` is. Also, if it is a list, then the variable name is not the best :wink:

File: tests/test_openjdk.py [hunk 3]
```diff
@@ -28,3 +30,39 @@ def test_jdk_version(container, java_version):
...
+    indirect=["container_per_test"]
+)
+@pytest.mark.parametrize(
+    "test_to_run, expected_string",
+    [
+        ("threads_concurrency_and_sleep" , ["I am the thread 1", "I am the thread 2"])
+    ],
+)
+def test_jdk_extended(
+    container_per_test, test_to_run, expected_string, host, container_runtime
```
Comment (dcermak) [inline tests/test_openjdk.py:49]: You can use type hints here:
```suggestion
    container_per_test, test_to_run: str, expected_string: List[str], host, container_runtime: OciRuntimeBase
```

File: tests/test_openjdk.py [hunk 4]
```diff
@@ -28,3 +30,39 @@ def test_jdk_version(container, java_version):
...
+    ],
+    indirect=["container_per_test"]
+)
+@pytest.mark.parametrize(
+    "test_to_run, expected_string",
+    [
+        ("threads_concurrency_and_sleep" , ["I am the thread 1", "I am the thread 2"])
+    ],
+)
+def test_jdk_extended(
```
Comment (dcermak) [inline tests/test_openjdk.py:48]: Please add some documentation for this test.

File: tests/trainers/java/threads_concurrency_and_sleep.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
+public class Test {
```
Comment (dcermak) [inline tests/trainers/java/threads_concurrency_and_sleep.java:1]: Please add some documentation for this file explaining what it does for those of us who have no clue about Java.

**Review by dcermak — CHANGES_REQUESTED** [review]
Looks good, but I'd appreciate if you'd document the test a bit more.

**Comment by dcermak** [comment]
@ilausuch Please format your code via `tox -e format` and rebase your changes on top of `main`.

**Comment by ilausuch** [comment]
Thanks Dan!

File: tests/test_openjdk.py (hunk 4 above)
Comment (ilausuch) [reply tests/test_openjdk.py:75]: Done

File: tests/test_openjdk.py (hunk 3 above)
Comment (ilausuch) [reply tests/test_openjdk.py]: Done. Thanks

File: tests/test_openjdk.py (hunk 2 above)
Comment (ilausuch) [reply tests/test_openjdk.py]: Done

File: tests/test_openjdk.py [hunk 6]
```diff
@@ -28,3 +56,38 @@ def test_jdk_version(container, java_version):
...
+        (
+            "threads_concurrency_and_sleep",
+            ["I am the thread 1", "I am the thread 2"],
+        )
+    ],
+)
+def test_jdk_extended(
+    container_per_test,
+    test_to_run: str,
+    expected_strings: list[str],
```
Comment (dcermak) [inline tests/test_openjdk.py:78]: Only python 3.8 and later support using the type `list` in type hints, for compatibility with python < 3.8 you have to resort to the `List` type from the `typing` module.
```suggestion
    expected_strings: List[str],
```

File: tests/trainers/java/threads_concurrency_and_sleep.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
+/**
+Test to check the Java Threading module.
+It runs two concurrent tests and wait for their ending.
+Also includes a pasive wait (sleep) in one of the thread.
```
Comment (dcermak) [inline tests/trainers/java/threads_concurrency_and_sleep.java:4]:
```suggestion
It runs two concurrent tests and waits for them to end.
The `BasicThread2` child class additionally includes a passive wait (sleep) for 1 second.
```

File: tests/test_openjdk.py [hunk 8]
```diff
@@ -28,3 +56,38 @@ def test_jdk_version(container, java_version):
...
+    """Executes a set of java tests stored on test/trainers/java.
+    The check is a list of expected strings.
+    """
+
+    # run the test
+    testout = container_per_test.connection.run_expect(
+        [0], "java " + condir + test_to_run + ".java"
+    )
+
+    # check the results
```
Comment (dcermak) [inline tests/test_openjdk.py:91]: I would suggest to not add comments that just duplicate the code

```suggestion
    testout = container_per_test.connection.run_expect(
        [0], "java " + condir + test_to_run + ".java"
    )

```

File: tests/test_openjdk.py [hunk 9]
```diff
@@ -28,3 +56,38 @@ def test_jdk_version(container, java_version):
...
+)
+def test_jdk_extended(
+    container_per_test,
+    test_to_run: str,
+    expected_strings: list[str],
+    host,
+    container_runtime: OciRuntimeBase,
+):
+    """Executes a set of java tests stored on test/trainers/java.
+    The check is a list of expected strings.
```
Comment (dcermak) [inline tests/test_openjdk.py:83]: This does not clarify what the test executes

File: tests/test_openjdk.py [hunk 10]
```diff
@@ -2,6 +2,34 @@
...
+CONTAINER_IMAGES_EXTENDED = [
+    pytest.param(
+        DerivedContainer(
+            base=container_from_pytest_param(CONTAINER_T),
+            containerfile=DOCKERF_EXTENDED,
+        ),
+        marks=CONTAINER_T.marks,
+        id=CONTAINER_T.id,
+    )
+    for CONTAINER_T in CONTAINER_IMAGES
```
Comment (dcermak) [inline tests/test_openjdk.py:31]: Usually only constants use all-caps
```suggestion
    pytest.param(
        DerivedContainer(
            base=container_from_pytest_param(container),
            containerfile=DOCKERF_EXTENDED,
        ),
        marks=container.marks,
        id=container.id,
    )
    for container in CONTAINER_IMAGES
```

**Review by dcermak — CHANGES_REQUESTED** [review]
We're getting there :wink:

File: tests/test_openjdk.py (hunk 8 above)
Comment (ilausuch) [reply tests/test_openjdk.py:91]: I agree, these comments came from when I had to read the code carefully to understand what was going on. removing...

File: tests/test_openjdk.py (hunk 10 above)
Comment (ilausuch) [reply tests/test_openjdk.py:31]: yes, I copy this from somewhere

Comment (ilausuch) [reply tests/test_openjdk.py]: done

File: tests/test_openjdk.py (hunk 9 above)
Comment (ilausuch) [reply tests/test_openjdk.py]: better now?

File: tests/test_openjdk.py (hunk 8 above)
Comment (ilausuch) [reply tests/test_openjdk.py]: done

File: tests/trainers/java/threads_concurrency_and_sleep.java (hunk 7 above)
Comment (ilausuch) [reply tests/trainers/java/threads_concurrency_and_sleep.java:4]: done, thanks

File: tests/test_openjdk.py (hunk 6 above)
Comment (ilausuch) [reply tests/test_openjdk.py]: done. I had to import List form typing

File: tests/trainers/java/threads_concurrency_and_sleep.java (hunk 5 above)
Comment (ilausuch) [reply tests/trainers/java/threads_concurrency_and_sleep.java:6]: done

File: tests/test_openjdk.py (hunk 4 above)
Comment (ilausuch) [reply tests/test_openjdk.py:76]: done

File: tests/trainers/java/threads_concurrency_and_sleep.java [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+    public static void main(String[] args) {
+        BasicThread t1 = new BasicThread("1");
+        t1.start();
+        BasicThread2 t2 = new BasicThread2("2");
+        t2.start();
+
+        try {
+          t1.join();
+          t2.join();
+        } catch ( Exception e) {
```
Comment (grisu48) [inline tests/trainers/java/threads_concurrency_and_sleep.java:16]: Catching `Exception` will always work but is disencouraged. One should only catch RuntimeExceptions or more specifically, the actual RuntimeException that are occurring 

```suggestion
        } catch ( InterruptedException e) {
```

I don't think this is really important to do, but the change is also very cheap and should not hurt IMHO :slightly_smiling_face:

File: tests/trainers/java/threads_concurrency_and_sleep.java [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+
+class BasicThread2 extends BasicThread {
+  BasicThread2(String id) {
+    super(id);
+  }
+
+  public void run() {
+    try{
+      Thread.sleep(1000);
+    }catch(Exception err){
```
Comment (grisu48) [inline tests/trainers/java/threads_concurrency_and_sleep.java:46]:
```suggestion
    }catch(InterruptedException err){
        System.out.println("Error in sleep: " + err.getMessage());
```

**Review by grisu48 — COMMENTED** [review]
Some nitpicking, otherwise LGTM. And it's nice to have the BCI Java tests being extended

File: tests/trainers/java/threads_concurrency_and_sleep.java (hunk 11 above)
Comment (ilausuch) [reply tests/trainers/java/threads_concurrency_and_sleep.java]: done

**Review by dcermak — APPROVED** [review]
LGTM, thanks!

**Pull request merged**